spec/json_validator_spec.rb

```diff
@@ -1,108 +1,133 @@
-      it { expect(user).to_not be_valid }
+    context 'with valid JSON data' do
+      let(:data) { 'What? This is not JSON at all.' }
+      it { expect(record.data_invalid_json).to eql(data) }
     end
 
-    context 'as JSON string' do
-      let(:attributes) { { name: 'Samuel Garneau', profile: '{ "city": "Quebec City" }' } }
-      it { expect(user).to_not be_valid }
+    context 'with invalid JSON data' do
+      let(:data) { { foo: 'bar' } }
+      it { expect(record.data_invalid_json).to be_nil }
     end
   end
 
-  context 'with valid JSON value' do
-    context 'as Ruby Hash' do
-      let(:attributes) { { name: 'Samuel Garneau', profile: { country: 'CA' } } }
-      it { expect(user).to be_valid }
+  describe :validate_each do
+    let(:validator) { JsonValidator.new(options) }
+    let(:options) { { attributes: [attribute], options: { strict: true } } }
+    let(:validate_each!) { validator.validate_each(record, attribute, value) }
+
+    # Doubles
+    let(:attribute) { double(:attribute, to_s: 'attribute_name') }
+    let(:record) { double(:record, errors: record_errors) }
+    let(:record_errors) { double(:errors) }
+    let(:value) { double(:value) }
+    let(:schema) { double(:schema) }
+    let(:validatable_value) { double(:validatable_value) }
+    let(:validator_errors) { double(:validator_errors) }
+
+    before do
+      expect(validator).to receive(:schema).with(record).and_return(schema)
+      expect(validator).to receive(:validatable_value).with(value).and_return(validatable_value)
+      expect(::JSON::Validator).to receive(:fully_validate).with(schema, validatable_value, options[:options]).and_return(validator_errors)
     end
 
-    context 'as JSON string' do
-      let(:attributes) { { name: 'Samuel Garneau', profile: '{ "country": "CA" }' } }
-      it { expect(user).to be_valid }
+    context 'with JSON::Validator errors' do
+      before do
+        expect(validator_errors).to receive(:empty?).and_return(false)
+        expect(record).not_to receive(:"#{attribute}_invalid_json")
+        expect(record_errors).to receive(:add).with(attribute, options[:message], value: value)
+      end
+
+      specify { validate_each! }
     end
-  end
 
-  context 'with malformed JSON string' do
-    let(:attributes) { { name: 'Samuel Garneau', profile: 'foo:}bar' } }
+    context 'without JSON::Validator errors but with invalid JSON data' do
+      before do
+        expect(validator_errors).to receive(:empty?).and_return(true)
+        expect(record).to receive(:"#{attribute}_invalid_json").and_return('foo"{]')
+        expect(record_errors).to receive(:add).with(attribute, options[:message], value: value)
+      end
 
-    specify do
-      expect(user).to_not be_valid
-      expect(user.profile).to eql({})
-      expect(user.profile_invalid_json).to eql('foo:}bar')
+      specify { validate_each! }
+    end
+
+    context 'without JSON::Validator errors and no invalid JSON data' do
```

Comment:
Picky, but I think `valid` is better than `no invalid` in this case (double negative). What do you think?